genie-core/src/main/java/com/netflix/genie/core/jpa/services/JpaBaseService.java

```diff
@@ -0,0 +1,181 @@
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.core.jpa.services;
+
+import com.google.common.collect.Sets;
+import com.netflix.genie.common.exceptions.GenieException;
+import com.netflix.genie.common.exceptions.GenieNotFoundException;
+import com.netflix.genie.common.exceptions.GenieServerException;
+import com.netflix.genie.core.jpa.entities.FileEntity;
+import com.netflix.genie.core.jpa.entities.TagEntity;
+import com.netflix.genie.core.jpa.repositories.JpaFileRepository;
+import com.netflix.genie.core.jpa.repositories.JpaTagRepository;
+import com.netflix.genie.core.services.FileService;
+import com.netflix.genie.core.services.TagService;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Base service for other services to extend for common functionality.
+ *
+ * @author tgianos
+ * @since 3.3.0
+ */
+@Slf4j
+@Getter(AccessLevel.PACKAGE)
+class JpaBaseService {
+
+    static final String GENIE_TAG_NAMESPACE = "genie.";
+    static final String GENIE_ID_TAG_NAMESPACE = GENIE_TAG_NAMESPACE + "id:";
+    static final String GENIE_NAME_TAG_NAMESPACE = GENIE_TAG_NAMESPACE + "name:";
+    private static final char COMMA = ',';
+    private static final String EMPTY_STRING = "";
+
+    private final TagService tagService;
+    private final JpaTagRepository tagRepository;
+    private final FileService fileService;
+    private final JpaFileRepository fileRepository;
+
+    /**
+     * Constructor.
+     *
+     * @param tagService          The tag service to use
+     * @param tagRepository       The tag repository to use
+     * @param fileService         The file service to use
+     * @param fileRepository      The file repository to use
+     */
+    JpaBaseService(
+        final TagService tagService,
+        final JpaTagRepository tagRepository,
+        final FileService fileService,
+        final JpaFileRepository fileRepository
+    ) {
+        this.tagService = tagService;
+        this.tagRepository = tagRepository;
+        this.fileService = fileService;
+        this.fileRepository = fileRepository;
+    }
+
+    /**
+     * Create a file reference in the database and return the attached entity.
+     *
+     * @param file The file to create
+     * @return The file entity that has been persisted in the database
+     * @throws GenieException on error
+     */
+    FileEntity createAndGetFileEntity(final String file) throws GenieException {
+        this.fileService.createFileIfNotExists(file);
```

Comment:
Isn't it better to return the FileEntity if exists or return the attached one after creation

Reply:
This was written when the `createFileIfNotExists` method still required a new transaction. So I needed to do the get of the attached entity once it returned to the current transaction context since the entity manager would be different for both transactions. Now that it's one transaction I probably could do as you say however leaving it as is may be safer if we have to revert to multiple transactions. Additionally since this `find` would use the current entity manager in the same transaction the entity should already be attached in the scope so it wouldn't go to the database again correct @ajoymajumdar ?

Reply:
Yes.